File: skills/code-refactoring/scripts/analyze_multidim.js
#!/usr/bin/env node
/**
 * Multi-Dimensional Code Analysis for JavaScript/TypeScript/Next.js Projects
 * Analyzes code across 5 dimensions: Maintainability, Performance, Security, Scalability, Reusability
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const { parseArgs } = require('node:util');

// Required packages for each dimension
const REQUIRED_PACKAGES = {
  maintainability: ['eslint', '@typescript-eslint/eslint-plugin'],
  performance: ['webpack-bundle-analyzer', 'lighthouse'],
  security: ['eslint-plugin-security', 'npm-audit'],
  scalability: ['dependency-cruiser', 'madge'],
  reusability: ['jscpd']
};

function checkPackageInstalled(packageName) {
  try {
    require.resolve(packageName);
    return true;
  } catch {
    return false;
  }
}

function checkDependencies(dimensions) {
  const available = {};
  
  for (const dim of dimensions) {
    if (!(dim in REQUIRED_PACKAGES)) continue;
    
    let dimAvailable = true;
    for (const pkg of REQUIRED_PACKAGES[dim]) {
      if (!checkPackageInstalled(pkg)) {
        console.log(`⚠️  ${pkg} not installed - ${dim} analysis may be limited`);
        dimAvailable = false;
      }
    }
    
    available[dim] = dimAvailable;
  }
  
  return available;
}

class MultiDimensionalAnalyzer {
  constructor(projectPath, dimensions) {
    this.projectPath = path.resolve(projectPath);
    this.dimensions = dimensions;
    this.results = {
      project: projectPath,
      dimensions: {},
      overall_health: 0,
      priority_actions: []
    };
  }

  async analyze() {
    console.log(`\n🔍 Analyzing ${this.projectPath} across ${this.dimensions.length} dimensions...\n`);
    
    if (this.dimensions.includes('maintainability')) {
      this.results.dimensions.maintainability = await this.analyzeMaintainability();
    }
    
    if (this.dimensions.includes('performance')) {
      this.results.dimensions.performance = await this.analyzePerformance();
    }
    
    if (this.dimensions.includes('security')) {
      this.results.dimensions.security = await this.analyzeSecurity();
    }
    
    if (this.dimensions.includes('scalability')) {
      this.results.dimensions.scalability = await this.analyzeScalability();
    }
    
    if (this.dimensions.includes('reusability')) {
      this.results.dimensions.reusability = await this.analyzeReusability();
    }
    
    // Calculate overall health
    this.calculateOverallHealth();
    
    // Generate priority actions
    this.generatePriorityActions();
    
    return this.results;
  }

  async analyzeMaintainability() {
    console.log('🔧 Analyzing Maintainability...');
    
    const result = {
      score: 0,
      coverage: 'comprehensive',
      metrics: {},
      issues: []
    };
    
    // Run ESLint for code quality
    try {
      const eslintPath = path.join(this.projectPath, 'node_modules', '.bin', 'eslint');
      
      if (fs.existsSync(eslintPath)) {
        const output = execSync(
          `${eslintPath} "${this.projectPath}" --format json --max-warnings 0 || true`,
          { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 }
        );
        
        const eslintResults = JSON.parse(output);
        
        let errorCount = 0;
        let warningCount = 0;
        const complexityIssues = [];
        
        for (const file of eslintResults) {
          errorCount += file.errorCount || 0;
          warningCount += file.warningCount || 0;
          
          // Look for complexity issues
          for (const message of file.messages || []) {
            if (message.ruleId === 'complexity') {
              complexityIssues.push({
                file: file.filePath,
                line: message.line,
                message: message.message,
                severity: message.severity === 2 ? 'error' : 'warning'
              });
            }
          }
        }
        
        result.metrics.eslint_errors = errorCount;
        result.metrics.eslint_warnings = warningCount;
        result.metrics.complexity_issues = complexityIssues.length;
        result.issues = complexityIssues;
        
        // Calculate score
        if (errorCount === 0 && warningCount < 10) {
          result.score = 95;
        } else if (errorCount < 5 && warningCount < 50) {
          result.score = 80;
        } else if (errorCount < 20) {
          result.score = 60;
        } else {
          result.score = 40;
        }
      }
    } catch (error) {
      console.log(`  Warning: ESLint analysis failed - ${error.message}`);
      result.score = 70; // Neutral score if can't analyze
    }
    
    console.log(`  ✓ Maintainability score: ${result.score}/100`);
    return result;
  }

  async analyzePerformance() {
    console.log('⚡ Analyzing Performance...');
    
    const result = {
      score: 70, // Default neutral score
      bottlenecks: [],
      improvements: [],
      bundle_issues: []
    };
    
    // Check for Next.js configuration
    const nextConfigPath = path.join(this.projectPath, 'next.config.js');
    const nextConfigMjsPath = path.join(this.projectPath, 'next.config.mjs');
    
    if (fs.existsSync(nextConfigPath) || fs.existsSync(nextConfigMjsPath)) {
      result.improvements.push({
        type: 'nextjs_optimization',
        message: 'Next.js project detected - check for Image optimization, bundle size',
        action: 'Use next/image, implement lazy loading, check bundle analyzer'
      });
    }
    
    // Static analysis for React performance issues
    const jsFiles = this.findFiles(this.projectPath, ['.js', '.jsx', '.ts', '.tsx']);
    
    let unnecessaryRerenderCount = 0;
    let missingSplittingCount = 0;
    
    for (const file of jsFiles) {
      try {
        const content = fs.readFileSync(file, 'utf-8');
        
        // Check for missing React.memo on components with props
        if (content.includes('export default function') || content.includes('export const')) {
          if (!content.includes('React.memo') && !content.includes('useMemo')) {
            unnecessaryRerenderCount++;
          }
        }
        
        // Check for large imports without lazy loading
        if (content.match(/import .+ from ['"]@\/components\/.*['"]/g)) {
          if (!content.includes('dynamic(') && !content.includes('lazy(')) {
            missingSplittingCount++;
          }
        }
      } catch (error) {
        continue;
      }
    }
    
    result.improvements.push({
      type: 'react_optimization',
      count: unnecessaryRerenderCount,
      message: `Found ${unnecessaryRerenderCount} components without memoization`
    });
    
    result.improvements.push({
      type: 'code_splitting',
      count: missingSplittingCount,
      message: `Found ${missingSplittingCount} files without lazy loading`
    });
    
    // Calculate score
    const totalIssues = unnecessaryRerenderCount + missingSplittingCount;
    if (totalIssues === 0) {
      result.score = 95;
    } else if (totalIssues < 10) {
      result.score = 80;
    } else if (totalIssues < 30) {
      result.score = 65;
    } else {
      result.score = 50;
    }
    
    console.log(`  ✓ Performance score: ${result.score}/100`);
    console.log(`  Found ${totalIssues} potential performance issues`);
    
    return result;
  }

  async analyzeSecurity() {
    console.log('🔒 Analyzing Security...');
    
    const result = {
      score: 100, // Start with perfect score
      vulnerabilities: [],
      severity: 'none'
    };
    
    // Run npm audit
    try {
      const auditOutput = execSync('npm audit --json', {
        cwd: this.projectPath,
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'ignore']
      });
      
      const auditData = JSON.parse(auditOutput);
      
      if (auditData.metadata) {
        const { vulnerabilities } = auditData.metadata;
        
        const criticalCount = vulnerabilities?.critical || 0;
        const highCount = vulnerabilities?.high || 0;
        const moderateCount = vulnerabilities?.moderate || 0;
        
        result.vulnerabilities.push({
          type: 'dependencies',
          critical: criticalCount,
          high: highCount,
          moderate: moderateCount,
          total: criticalCount + highCount + moderateCount
        });
        
        // Calculate score
        if (criticalCount > 0) {
          result.severity = 'critical';
          result.score = Math.max(20, 90 - (criticalCount * 20));
        } else if (highCount > 0) {
          result.severity = 'high';
          result.score = Math.max(40, 90 - (highCount * 10));
        } else if (moderateCount > 0) {
          result.severity = 'moderate';
          result.score = Math.max(70, 90 - (moderateCount * 2));
        }
      }
    } catch (error) {
      // npm audit returns non-zero when vulnerabilities found
      if (error.stdout) {
        try {
          const auditData = JSON.parse(error.stdout);
          // Process similar to above
        } catch {}
      }
    }
    
    // Static analysis for security patterns
    const jsFiles = this.findFiles(this.projectPath, ['.js', '.jsx', '.ts', '.tsx']);
    
    const securityIssues = [];
    for (const file of jsFiles) {
      try {
        const content = fs.readFileSync(file, 'utf-8');
        
        // Check for common security issues
        if (content.includes('dangerouslySetInnerHTML')) {
          securityIssues.push({
            file,
            type: 'xss_risk',
            message: 'Using dangerouslySetInnerHTML - XSS risk'
          });
        }
        
        if (content.match(/eval\(/)) {
          securityIssues.push({
            file,
            type: 'code_injection',
            message: 'Using eval() - code injection risk'
          });
        }
        
        // Check for hardcoded secrets (basic pattern)
        if (content.match(/(apiKey|api_key|secret|password|token)\s*=\s*['"][^'"]{20,}['"]/i)) {
          securityIssues.push({
            file,
            type: 'hardcoded_secret',
            message: 'Potential hardcoded secret detected'
          });
        }
      } catch (error) {
        continue;
      }
    }
    
    if (securityIssues.length > 0) {
      result.vulnerabilities.push({
        type: 'code_patterns',
        issues: securityIssues,
        count: securityIssues.length
      });
      
      result.score = Math.min(result.score, 80 - (securityIssues.length * 5));
    }
    
    console.log(`  ✓ Security score: ${result.score}/100`);
    console.log(`  Found ${result.vulnerabilities.length} security issue categories`);
    
    return result;
  }

  async analyzeScalability() {
    console.log('📈 Analyzing Scalability...');
    
    const result = {
      score: 75, // Default neutral score
      coupling_issues: [],
      circular_dependencies: [],
      architecture_issues: []
    };
    
    // Check for circular dependencies using madge
    try {
      const madgePath = path.join(this.projectPath, 'node_modules', '.bin', 'madge');
      
      if (fs.existsSync(madgePath)) {
        const output = execSync(
          `${madgePath} --circular --json "${this.projectPath}"`,
          { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 }
        );
        
        const circularDeps = JSON.parse(output);
        
        for (const [file, deps] of Object.entries(circularDeps)) {
          if (deps.length > 0) {
            result.circular_dependencies.push({
              file,
              dependencies: deps
            });
          }
        }
      }
    } catch (error) {
      // Madge not available or no circular deps
    }
    
    // Check for large components (god components)
    const componentFiles = this.findFiles(this.projectPath, ['.jsx', '.tsx']);
    
    for (const file of componentFiles) {
      try {
        const content = fs.readFileSync(file, 'utf-8');
        const lines = content.split('\n').length;
        
        if (lines > 300) {
          result.architecture_issues.push({
            file,
            lines,
            type: 'god_component',
            message: `Component has ${lines} lines - consider splitting`
          });
        }
      } catch (error) {
        continue;
      }
    }
    
    // Calculate score
    const issueCount = result.circular_dependencies.length + result.architecture_issues.length;
    
    if (issueCount === 0) {
      result.score = 90;
    } else if (issueCount < 5) {
      result.score = 75;
    } else if (issueCount < 10) {
      result.score = 60;
    } else {
      result.score = 40;
    }
    
    console.log(`  ✓ Scalability score: ${result.score}/100`);
    console.log(`  Found ${issueCount} scalability issues`);
    
    return result;
  }

  async analyzeReusability() {
    console.log('♻️  Analyzing Reusability...');
    
    const result = {
      score: 80, // Default good score
      duplication_percentage: 0,
      duplicate_blocks: [],
      extractable_patterns: []
    };
    
    // Check for duplicate code using jscpd
    try {
      const jscpdPath = path.join(this.projectPath, 'node_modules', '.bin', 'jscpd');
      
      if (fs.existsSync(jscpdPath)) {
        const output = execSync(
          `${jscpdPath} "${this.projectPath}" --format json --silent || true`,
          { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 }
        );
        
        if (output) {
          const jscpdData = JSON.parse(output);
          
          if (jscpdData.statistics) {
            result.duplication_percentage = jscpdData.statistics.percentage || 0;
            result.duplicate_blocks = jscpdData.duplicates?.length || 0;
            
            // Calculate score based on duplication
            if (result.duplication_percentage < 3) {
              result.score = 95;
            } else if (result.duplication_percentage < 5) {
              result.score = 85;
            } else if (result.duplication_percentage < 10) {
              result.score = 70;
            } else {
              result.score = 50;
            }
          }
        }
      }
    } catch (error) {
      console.log(`  Warning: jscpd analysis failed - ${error.message}`);
    }
    
    console.log(`  ✓ Reusability score: ${result.score}/100`);
    console.log(`  Code duplication: ${result.duplication_percentage.toFixed(2)}%`);
    
    return result;
  }

  findFiles(dir, extensions) {
    const files = [];
    
    const walk = (currentDir) => {
      try {
        const entries = fs.readdirSync(currentDir, { withFileTypes: true });
        
        for (const entry of entries) {
          // Skip node_modules and hidden directories
          if (entry.name === 'node_modules' || entry.name.startsWith('.')) {
            continue;
          }
          
          const fullPath = path.join(currentDir, entry.name);
          
          if (entry.isDirectory()) {
            walk(fullPath);
          } else if (extensions.some(ext => entry.name.endsWith(ext))) {
            files.push(fullPath);
          }
        }
      } catch (error) {
        // Permission denied or other errors
      }
    };
    
    walk(dir);
    return files;
  }

  calculateOverallHealth() {
    const scores = [];
    const weights = {
      maintainability: 0.35,
      performance: 0.20,
      security: 0.25,
      scalability: 0.10,
      reusability: 0.10
    };
    
    for (const [dim, weight] of Object.entries(weights)) {
      if (this.results.dimensions[dim]) {
        const dimScore = this.results.dimensions[dim].score || 70;
        scores.push(dimScore * weight);
      }
    }
    
    this.results.overall_health = Math.round(scores.reduce((a, b) => a + b, 0));
  }

  generatePriorityActions() {
    const actions = [];
    
    for (const [dim, data] of Object.entries(this.results.dimensions)) {
      const score = data.score || 0;
      
      if (score < 60) {
        actions.push({
          dimension: dim,
          priority: 'high',
          score,
          action: `Address critical ${dim} issues immediately`
        });
      } else if (score < 75) {
        actions.push({
          dimension: dim,
          priority: 'medium',
          score,
          action: `Improve ${dim} in next iteration`
        });
      }
    }
    
    // Sort by score (lowest first)
    actions.sort((a, b) => a.score - b.score);
    
    this.results.priority_actions = actions.slice(0, 5);
  }
}

async function main() {
  const { values, positionals } = parseArgs({
    options: {
      dimensions: {
        type: 'string',
        default: 'all'
      },
      output: {
        type: 'string',
        default: '/mnt/user-data/outputs/multidim-analysis.json'
      },
      all: {
        type: 'boolean',
        default: false
      }
    },
    allowPositionals: true
  });
  
  if (positionals.length === 0) {
    console.error('Usage: node analyze_multidim.js <project-path> [options]');
    process.exit(1);
  }
  
  const projectPath = positionals[0];
  
  // Parse dimensions
  let dimensions;
  if (values.all || values.dimensions === 'all') {
    dimensions = ['maintainability', 'performance', 'security', 'scalability', 'reusability'];
  } else {
    dimensions = values.dimensions.split(',').map(d => d.trim());
  }
  
  // Check dependencies
  console.log('\n🔍 Checking analysis tools...');
  checkDependencies(dimensions);
  
  // Run analysis
  const analyzer = new MultiDimensionalAnalyzer(projectPath, dimensions);
  const results = await analyzer.analyze();
  
  // Save results
  const outputDir = path.dirname(values.output);
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }
  
  fs.writeFileSync(values.output, JSON.stringify(results, null, 2));
  
  console.log('\n✅ Analysis complete!');
  console.log(`📊 Overall Health Score: ${results.overall_health}/100`);
  console.log(`📄 Full report: ${values.output}`);
  
  // Print priority actions
  if (results.priority_actions.length > 0) {
    console.log('\n🎯 Priority Actions:');
    for (const action of results.priority_actions) {
      console.log(`  [${action.priority.toUpperCase()}] ${action.action} (score: ${action.score})`);
    }
  }
  
  process.exit(0);
}

main().catch(error => {
  console.error('Error:', error);
  process.exit(1);
});

File: skills/code-refactoring/scripts/analyze_multidim.py
#!/usr/bin/env python3
"""
Multi-Dimensional Code Analysis for Python Projects
Analyzes code across 5 dimensions: Maintainability, Performance, Security, Scalability, Reusability
"""

import argparse
import json
import os
import subprocess
import sys
from pathlib import Path
from typing import Dict, List, Any, Optional
import ast
import importlib.util

# Check for required packages
REQUIRED_PACKAGES = {
    'maintainability': ['radon', 'pylint', 'rope'],
    'performance': ['line_profiler', 'memory_profiler'],
    'security': ['bandit', 'safety'],
    'scalability': ['pydeps'],
    'reusability': ['pylint']  # Uses similarity checker
}

def check_package_installed(package: str) -> bool:
    """Check if a Python package is installed"""
    spec = importlib.util.find_spec(package.replace('-', '_'))
    return spec is not None

def check_dependencies(dimensions: List[str]) -> Dict[str, bool]:
    """Check which analysis tools are available"""
    available = {}
    
    for dim in dimensions:
        if dim not in REQUIRED_PACKAGES:
            continue
        
        dim_available = True
        for pkg in REQUIRED_PACKAGES[dim]:
            if not check_package_installed(pkg):
                print(f"⚠️  {pkg} not installed - {dim} analysis may be limited")
                dim_available = False
        
        available[dim] = dim_available
    
    return available


class MultiDimensionalAnalyzer:
    """Analyzes Python code across multiple dimensions"""
    
    def __init__(self, project_path: str, dimensions: List[str]):
        self.project_path = Path(project_path)
        self.dimensions = dimensions
        self.results = {
            'project': str(project_path),
            'dimensions': {},
            'overall_health': 0,
            'priority_actions': []
        }
    
    def analyze(self) -> Dict[str, Any]:
        """Run analysis for all requested dimensions"""
        print(f"\n🔍 Analyzing {self.project_path} across {len(self.dimensions)} dimensions...\n")
        
        if 'maintainability' in self.dimensions:
            self.results['dimensions']['maintainability'] = self.analyze_maintainability()
        
        if 'performance' in self.dimensions:
            self.results['dimensions']['performance'] = self.analyze_performance()
        
        if 'security' in self.dimensions:
            self.results['dimensions']['security'] = self.analyze_security()
        
        if 'scalability' in self.dimensions:
            self.results['dimensions']['scalability'] = self.analyze_scalability()
        
        if 'reusability' in self.dimensions:
            self.results['dimensions']['reusability'] = self.analyze_reusability()
        
        # Calculate overall health
        self._calculate_overall_health()
        
        # Generate priority actions
        self._generate_priority_actions()
        
        return self.results
    
    def analyze_maintainability(self) -> Dict[str, Any]:
        """Analyze maintainability dimension"""
        print("🔧 Analyzing Maintainability...")
        
        result = {
            'score': 0,
            'coverage': 'comprehensive',
            'metrics': {},
            'issues': []
        }
        
        # Run Radon for complexity
        try:
            radon_output = subprocess.run(
                ['radon', 'cc', str(self.project_path), '-a', '--json'],
                capture_output=True,
                text=True,
                timeout=30
            )
            
            if radon_output.returncode == 0:
                complexity_data = json.loads(radon_output.stdout)
                
                # Calculate average complexity
                total_complexity = 0
                func_count = 0
                high_complexity = []
                
                for file, functions in complexity_data.items():
                    for func in functions:
                        total_complexity += func.get('complexity', 0)
                        func_count += 1
                        
                        if func.get('complexity', 0) > 10:
                            high_complexity.append({
                                'file': file,
                                'function': func.get('name'),
                                'complexity': func.get('complexity'),
                                'line': func.get('lineno')
                            })
                
                avg_complexity = total_complexity / func_count if func_count > 0 else 0
                result['metrics']['average_complexity'] = round(avg_complexity, 2)
                result['metrics']['high_complexity_functions'] = len(high_complexity)
                result['issues'].extend(high_complexity)
                
                # Score based on complexity
                if avg_complexity <= 5:
                    result['score'] = 100
                elif avg_complexity <= 10:
                    result['score'] = 80
                elif avg_complexity <= 15:
                    result['score'] = 60
                else:
                    result['score'] = 40
                    
        except Exception as e:
            print(f"  Warning: Radon analysis failed - {e}")
        
        # Run Radon for maintainability index
        try:
            radon_mi = subprocess.run(
                ['radon', 'mi', str(self.project_path), '--json'],
                capture_output=True,
                text=True,
                timeout=30
            )
            
            if radon_mi.returncode == 0:
                mi_data = json.loads(radon_mi.stdout)
                
                mi_scores = []
                for file, data in mi_data.items():
                    mi_scores.append(data.get('mi', 0))
                
                avg_mi = sum(mi_scores) / len(mi_scores) if mi_scores else 0
                result['metrics']['maintainability_index'] = round(avg_mi, 2)
                
        except Exception as e:
            print(f"  Warning: MI calculation failed - {e}")
        
        print(f"  ✓ Maintainability score: {result['score']}/100")
        return result
    
    def analyze_performance(self) -> Dict[str, Any]:
        """Analyze performance dimension"""
        print("⚡ Analyzing Performance...")
        
        result = {
            'score': 70,  # Default neutral score
            'bottlenecks': [],
            'improvements': [],
            'algorithmic_issues': []
        }
        
        # Static analysis for algorithmic complexity
        python_files = list(self.project_path.rglob('*.py'))
        
        for py_file in python_files:
            try:
                with open(py_file, 'r', encoding='utf-8') as f:
                    tree = ast.parse(f.read())
                
                # Look for nested loops (potential O(n²) or worse)
                nested_loops = self._find_nested_loops(tree, str(py_file))
                result['algorithmic_issues'].extend(nested_loops)
                
            except Exception as e:
                continue
        
        # Score based on algorithmic issues
        if len(result['algorithmic_issues']) == 0:
            result['score'] = 90
        elif len(result['algorithmic_issues']) < 5:
            result['score'] = 70
        else:
            result['score'] = 50
        
        result['improvements'].append({
            'type': 'static_analysis',
            'message': 'Use profiling tools (cProfile, memory_profiler) for runtime analysis',
            'action': 'Install: pip install line-profiler memory-profiler'
        })
        
        print(f"  ✓ Performance score: {result['score']}/100")
        print(f"  Found {len(result['algorithmic_issues'])} potential algorithmic issues")
        
        return result
    
    def _find_nested_loops(self, tree: ast.AST, filename: str) -> List[Dict]:
        """Find nested loops that may indicate O(n²) complexity"""
        issues = []
        
        for node in ast.walk(tree):
            if isinstance(node, (ast.For, ast.While)):
                # Check if there's a nested loop
                for child in ast.walk(node):
                    if child != node and isinstance(child, (ast.For, ast.While)):
                        issues.append({
                            'file': filename,
                            'line': node.lineno,
                            'type': 'nested_loop',
                            'message': 'Nested loop detected - potential O(n²) complexity',
                            'severity': 'medium'
                        })
                        break
        
        return issues
    
    def analyze_security(self) -> Dict[str, Any]:
        """Analyze security dimension"""
        print("🔒 Analyzing Security...")
        
        result = {
            'score': 100,  # Start with perfect score
            'vulnerabilities': [],
            'severity': 'none'
        }
        
        # Run Bandit for security issues
        try:
            bandit_output = subprocess.run(
                ['bandit', '-r', str(self.project_path), '-f', 'json'],
                capture_output=True,
                text=True,
                timeout=60
            )
            
            # Bandit returns non-zero when it finds issues
            if bandit_output.stdout:
                bandit_data = json.loads(bandit_output.stdout)
                
                vulnerabilities = bandit_data.get('results', [])
                result['vulnerabilities'] = [
                    {
                        'file': v.get('filename'),
                        'line': v.get('line_number'),
                        'severity': v.get('issue_severity'),
                        'confidence': v.get('issue_confidence'),
                        'message': v.get('issue_text'),
                        'cwe': v.get('issue_cwe', {}).get('id')
                    }
                    for v in vulnerabilities
                ]
                
                # Calculate severity
                high_severity = sum(1 for v in result['vulnerabilities'] if v['severity'] == 'HIGH')
                medium_severity = sum(1 for v in result['vulnerabilities'] if v['severity'] == 'MEDIUM')
                
                if high_severity > 0:
                    result['severity'] = 'high'
                    result['score'] = max(30, 90 - (high_severity * 15))
                elif medium_severity > 0:
                    result['severity'] = 'medium'
                    result['score'] = max(50, 90 - (medium_severity * 5))
                else:
                    result['severity'] = 'low'
                    result['score'] = 85
                
        except FileNotFoundError:
            result['score'] = 0
            result['vulnerabilities'].append({
                'message': 'Bandit not installed - install with: pip install bandit --break-system-packages',
                'severity': 'INFO'
            })
        except Exception as e:
            print(f"  Warning: Bandit analysis failed - {e}")
        
        # Check for safety (dependency vulnerabilities)
        try:
            safety_output = subprocess.run(
                ['safety', 'check', '--json'],
                capture_output=True,
                text=True,
                timeout=30,
                cwd=str(self.project_path)
            )
            
            if safety_output.stdout:
                safety_data = json.loads(safety_output.stdout)
                if safety_data:
                    result['vulnerabilities'].append({
                        'type': 'dependency',
                        'count': len(safety_data),
                        'message': f'Found {len(safety_data)} vulnerable dependencies'
                    })
                    result['score'] = min(result['score'], 70)
                    
        except FileNotFoundError:
            pass  # Safety not installed, skip
        except Exception as e:
            print(f"  Warning: Safety check failed - {e}")
        
        print(f"  ✓ Security score: {result['score']}/100")
        print(f"  Found {len(result['vulnerabilities'])} security issues")
        
        return result
    
    def analyze_scalability(self) -> Dict[str, Any]:
        """Analyze scalability/extensibility dimension"""
        print("📈 Analyzing Scalability...")
        
        result = {
            'score': 75,  # Default neutral score
            'coupling_issues': [],
            'solid_violations': [],
            'circular_dependencies': []
        }
        
        # Analyze imports and dependencies
        python_files = list(self.project_path.rglob('*.py'))
        
        import_graph = {}
        for py_file in python_files:
            try:
                with open(py_file, 'r', encoding='utf-8') as f:
                    tree = ast.parse(f.read())
                
                imports = []
                for node in ast.walk(tree):
                    if isinstance(node, ast.Import):
                        for alias in node.names:
                            imports.append(alias.name)
                    elif isinstance(node, ast.ImportFrom):
                        if node.module:
                            imports.append(node.module)
                
                import_graph[str(py_file)] = imports
                
            except Exception:
                continue
        
        # Detect god classes (classes with too many methods)
        god_classes = self._find_god_classes(python_files)
        result['solid_violations'].extend(god_classes)
        
        # Score based on violations
        violation_count = len(result['solid_violations'])
        if violation_count == 0:
            result['score'] = 90
        elif violation_count < 5:
            result['score'] = 75
        elif violation_count < 10:
            result['score'] = 60
        else:
            result['score'] = 40
        
        print(f"  ✓ Scalability score: {result['score']}/100")
        print(f"  Found {violation_count} SOLID violations")
        
        return result
    
    def _find_god_classes(self, python_files: List[Path]) -> List[Dict]:
        """Find classes with too many methods (God Class anti-pattern)"""
        god_classes = []
        
        for py_file in python_files:
            try:
                with open(py_file, 'r', encoding='utf-8') as f:
                    tree = ast.parse(f.read())
                
                for node in ast.walk(tree):
                    if isinstance(node, ast.ClassDef):
                        # Count methods
                        method_count = sum(1 for item in node.body if isinstance(item, ast.FunctionDef))
                        
                        if method_count > 20:  # Threshold for god class
                            god_classes.append({
                                'file': str(py_file),
                                'class': node.name,
                                'line': node.lineno,
                                'method_count': method_count,
                                'violation': 'Single Responsibility Principle',
                                'message': f'Class has {method_count} methods - consider splitting'
                            })
                
            except Exception:
                continue
        
        return god_classes
    
    def analyze_reusability(self) -> Dict[str, Any]:
        """Analyze reusability dimension"""
        print("♻️  Analyzing Reusability...")
        
        result = {
            'score': 80,  # Default good score
            'duplication_percentage': 0,
            'duplicate_blocks': [],
            'extractable_patterns': []
        }
        
        # Run Pylint similarity checker
        try:
            pylint_output = subprocess.run(
                ['pylint', '--disable=all', '--enable=similarities', 
                 str(self.project_path), '--output-format=json'],
                capture_output=True,
                text=True,
                timeout=60
            )
            
            if pylint_output.stdout:
                try:
                    pylint_data = json.loads(pylint_output.stdout)
                    
                    similar_code = [msg for msg in pylint_data if msg.get('symbol') == 'duplicate-code']
                    result['duplicate_blocks'] = [
                        {
                            'file': msg.get('path'),
                            'line': msg.get('line'),
                            'message': msg.get('message')
                        }
                        for msg in similar_code
                    ]
                    
                    # Estimate duplication percentage
                    if len(result['duplicate_blocks']) > 0:
                        result['duplication_percentage'] = min(len(result['duplicate_blocks']) * 2, 50)
                        result['score'] = max(50, 100 - (len(result['duplicate_blocks']) * 5))
                    
                except json.JSONDecodeError:
                    pass
                    
        except Exception as e:
            print(f"  Warning: Similarity analysis failed - {e}")
        
        print(f"  ✓ Reusability score: {result['score']}/100")
        print(f"  Found {len(result['duplicate_blocks'])} duplicate code blocks")
        
        return result
    
    def _calculate_overall_health(self):
        """Calculate overall health score from dimension scores"""
        scores = []
        weights = {
            'maintainability': 0.35,
            'performance': 0.20,
            'security': 0.25,
            'scalability': 0.10,
            'reusability': 0.10
        }
        
        for dim, weight in weights.items():
            if dim in self.results['dimensions']:
                dim_score = self.results['dimensions'][dim].get('score', 70)
                scores.append(dim_score * weight)
        
        self.results['overall_health'] = round(sum(scores))
    
    def _generate_priority_actions(self):
        """Generate prioritized list of actions"""
        actions = []
        
        for dim, data in self.results['dimensions'].items():
            score = data.get('score', 0)
            
            if score < 60:
                actions.append({
                    'dimension': dim,
                    'priority': 'high',
                    'score': score,
                    'action': f'Address critical {dim} issues immediately'
                })
            elif score < 75:
                actions.append({
                    'dimension': dim,
                    'priority': 'medium',
                    'score': score,
                    'action': f'Improve {dim} in next iteration'
                })
        
        # Sort by score (lowest first)
        actions.sort(key=lambda x: x['score'])
        
        self.results['priority_actions'] = actions[:5]  # Top 5 priorities


def main():
    parser = argparse.ArgumentParser(
        description='Multi-dimensional code analysis for Python projects'
    )
    parser.add_argument('project', help='Path to Python project')
    parser.add_argument(
        '--dimensions',
        default='all',
        help='Comma-separated dimensions: maintainability,performance,security,scalability,reusability (default: all)'
    )
    parser.add_argument(
        '--output',
        default='/mnt/user-data/outputs/multidim-analysis.json',
        help='Output file path'
    )
    parser.add_argument(
        '--all',
        action='store_true',
        help='Analyze all dimensions (same as --dimensions all)'
    )
    
    args = parser.parse_args()
    
    # Parse dimensions
    if args.all or args.dimensions == 'all':
        dimensions = ['maintainability', 'performance', 'security', 'scalability', 'reusability']
    else:
        dimensions = [d.strip() for d in args.dimensions.split(',')]
    
    # Check dependencies
    print("\n🔍 Checking analysis tools...")
    available = check_dependencies(dimensions)
    
    # Run analysis
    analyzer = MultiDimensionalAnalyzer(args.project, dimensions)
    results = analyzer.analyze()
    
    # Save results
    output_path = Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    
    with open(output_path, 'w') as f:
        json.dump(results, f, indent=2)
    
    print(f"\n✅ Analysis complete!")
    print(f"📊 Overall Health Score: {results['overall_health']}/100")
    print(f"📄 Full report: {output_path}")
    
    # Print priority actions
    if results['priority_actions']:
        print("\n🎯 Priority Actions:")
        for action in results['priority_actions']:
            print(f"  [{action['priority'].upper()}] {action['action']} (score: {action['score']})")
    
    return 0


if __name__ == '__main__':
    sys.exit(main())
